torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1455,6 +1508,20 @@ def _unshard(
         Postcondition: Each handle's ``FlatParameter`` 's data is the padded
         unsharded flattened parameter on the compute device.
         """
+        if self.limit_all_gathers:
+            queue = self._free_event_queue
+            events: List[FreeEvent] = []
+            if queue.queue_size > self._max_inflight_all_gather_size:
+                # Upon reaching the limit, flush the queue
+                while queue.queue_size > 0:
+                    event = queue.dequeue()
+                    if event is None:
+                        queue.check_integrity()
+                        assert 0, "Queue integrity check should have failed"
+                    events.append(event)
+            # As an optimization, only synchronize the latest event
+            if events:
+                events[-1].cuda_event.synchronize()
```

Comment:
just double confirm, when we synchronize the latest event, the previous events are synced, right? because this is a full CPU/GPU sync

Reply:
We may want some ask some experts to confirm this.

However, conceptually, it made sense to me that we only need to sync the latest event. By calling `.synchronize()`, we wait for that latest free to finish, at which point all earlier frees (which happen in the same stream, the default stream) must have finished as well since they are obey the sequential order ensured by the stream semantics.

I have not tested if there is any performance gain by only calling `.synchronize()` on the latest event. However, it seems that it cannot be slower compared to calling it on all of them.